dive_core/command_hierarchy.cpp

```diff
@@ -2089,32 +2093,157 @@ void CommandHierarchyCreator::AppendPacketFieldNodes(const IMemoryManager &mem_m
+                               submit_index,                                 \
+                               ubo_addr,                                     \
+                               sizeof(ubo_packet) / sizeof(uint32_t),        \
+                               false,                                        \
+                               packet_info_ptr,                              \
+                               packet_node_index,                            \
+                               prefix_stream.str().c_str());                 \
+    }
+
+    if (cat == StateBlockCat::kTex)
+    {
+        if (packet.bitfields0.STATE_TYPE == ST6_SHADER)
+            APPEND_SHARPS("A6XX_TEX_SAMP")
+        else if (packet.bitfields0.STATE_TYPE == ST6_CONSTANTS)
+            APPEND_SHARPS("A6XX_TEX_CONST")
+        else if (packet.bitfields0.STATE_TYPE == ST6_UBO)
+            APPEND_SHARPS("A6XX_UBO")
+    }
+    else if (cat == StateBlockCat::kShader)
+    {
+        // Shader program
+        if (packet.bitfields0.STATE_TYPE == ST6_SHADER)
+        {
+        }
+        // Constant data
+        else if (packet.bitfields0.STATE_TYPE == ST6_CONSTANTS)
+        {
+            // NUM_UNIT is in unit of float4s
+            // Add 4 dwords per line
+            for (uint32_t i = 0; i < packet.bitfields0.NUM_UNIT; ++i)
+            {
+                std::ostringstream string_stream;
+                for (uint32_t j = 0; j < 4; ++j)
+                {
+                    float    value;
+                    uint64_t addr = ext_src_addr + ((i * 4 + j) * sizeof(float));
+                    DIVE_VERIFY(mem_manager.CopyMemory(&value, submit_index, addr, sizeof(float)));
+                    string_stream << std::fixed << std::setw(11) << std::setprecision(6);
+                    string_stream << value << " ";
+                }
+
+                // Add it as child to packet_node
+                CommandHierarchy::AuxInfo aux_info = CommandHierarchy::AuxInfo::RegFieldNode(false);
+                uint64_t                  const_node_index = AddNode(NodeType::kFieldNode,
+                                                    string_stream.str(),
+                                                    aux_info);
+                AddChild(CommandHierarchy::kEngineTopology, packet_node_index, const_node_index);
+                AddChild(CommandHierarchy::kSubmitTopology, packet_node_index, const_node_index);
+                AddChild(CommandHierarchy::kAllEventTopology, packet_node_index, const_node_index);
+                AddChild(CommandHierarchy::kRgpTopology, packet_node_index, const_node_index);
+            }
+        }
+        else if (packet.bitfields0.STATE_TYPE == ST6_UBO)
+            APPEND_SHARPS("A6XX_UBO")
+        else if (packet.bitfields0.STATE_TYPE == ST6_IBO)
+        {
+            DIVE_ASSERT(packet.bitfields0.STATE_BLOCK == SB6_CS_SHADER);
+            APPEND_SHARPS("A6XX_TEX_CONST")
```

Comment:
why do we append A6XX_TEX_CONST when the state type is IBO? by the way, what is A6XX_TEX_CONST? texture constant?

Reply:
A6XX_TEX_CONST is a T#. IBO probably stands for Image Buffer Object?
I'm following cffdec.c's logic, and my reading of the code shows that IBO is only used when state-block is set to SB6_CS_SHADER. Here's the relevant code snippets:

`
      // SB4_CS_SHADER:
      [0xd][0] = {MESA_SHADER_COMPUTE, SHADER_PROG},
      [0xd][1] = {MESA_SHADER_COMPUTE, SHADER_CONST},
      [0xd][2] = {MESA_SHADER_COMPUTE, UBO},
      [0xd][3] = {MESA_SHADER_COMPUTE, SSBO_0}, /* a6xx location */
  ...
   case SSBO_0: {
      uint32_t *ssboconst = (uint32_t *)contents;

      for (i = 0; i < num_unit; i++) {
         int sz = 4;
         if (options->info->chip == 4) {
            dump_domain(ssboconst, 4, level + 2, "A4XX_SSBO_0");
         } else if (options->info->chip == 5) {
            dump_domain(ssboconst, 4, level + 2, "A5XX_SSBO_0");
         } else if ((6 <= options->info->chip) && (options->info->chip < 8)) {
            sz = 16;
            dump_domain(ssboconst, 16, level + 2, "A6XX_TEX_CONST");
         }
         dump_hex(ssboconst, sz, level + 1);
         ssboconst += sz;
      }
      break;
   }
`
The idea is that there's a lookup table, where [0xd][3] is where SB6_CS_SHADER=0xd and ST6_IBO=0x3. The case statements shows how it's used, and notice for 6xx and 7xx the SSBOs are in the form of T#s.

Reply:
good to know, thanks!